---
title: "Vignette"
author: "Dan Bunis"
date: "7/18/2019"
output: html_document
---

```{r, echo=FALSE}
knitr::opts_chunk$set(echo=TRUE, error=FALSE, warning=FALSE, message=FALSE)
```

## Download Raw PBMC data fomr 10X website

```{r loading}
# BiocManager::install("BiocFileCache")
library(BiocFileCache)
bfc <- BiocFileCache("raw_data", ask = FALSE)
raw.path <- bfcrpath(bfc, file.path("http://cf.10xgenomics.com/samples",
    "cell-exp/2.1.0/pbmc4k/pbmc4k_raw_gene_bc_matrices.tar.gz"))
untar(raw.path, exdir=file.path(tempdir(), "pbmc4k"))
# BiocManager::install("DropletUtils")
library(DropletUtils)
fname <- file.path(tempdir(), "pbmc4k/raw_gene_bc_matrices/GRCh38")
sce.pbmc <- read10xCounts(fname, col.names=TRUE)
```

```{r Exploration1}
# devtools::install_github("dtm2451/DittoSeq")
library(DittoSeq)
sce.pbmc
str(sce.pbmc, max.level = 3)
get.metas("sce.pbmc")
```

## Setup

#### Annotate genes with SYMBOL names rather than the Ensembl Gene ID thwy start as, and add mapping information

```{r gene-annotation}
# BiocManager::install("scater")
library(scater)
rownames(sce.pbmc) <- uniquifyFeatureNames(rowData(sce.pbmc)$ID, rowData(sce.pbmc)$Symbol)
# BiocManager::install("EnsDb.Hsapiens.v86")
library(EnsDb.Hsapiens.v86)
location <- mapIds(EnsDb.Hsapiens.v86, keys=rowData(sce.pbmc)$ID, 
    column="SEQNAME", keytype="GENEID")
```

```{r Exploration2}
sce.pbmc
```

#### Determine Cells from empty droplets

```{r cell-detection}
set.seed(100)
e.out <- emptyDrops(counts(sce.pbmc))
sce.pbmc <- sce.pbmc[,which(e.out$FDR <= 0.001)]
```

```{r Exploration3}
sce.pbmc
```

Notice that now there are only 4233 columns / cells in the object

#### Determine useable / live / high quality cells

```{r quality-control}
# Calculate QC statistics
sce.pbmc <- calculateQCMetrics(sce.pbmc, feature_controls=list(Mito=which(location=="MT")))
  sce.pbmc_hold <- sce.pbmc
# Determine outliers for total reads percent.mito per cell (makes a logical the length of our #cells)
high.mito <- isOutlier(sce.pbmc$pct_counts_Mito, nmads=3, type="higher")
# Remove cells with high mitochondrial read percentages
sce.pbmc <- sce.pbmc[,!high.mito]
```

```{r Exploration4}
sce.pbmc
```

We removed 311 cells with this metric and now have 3922 cells

#### Normalize cells

```{r normalization}
BiocManager::install("scran")
library(scran)
set.seed(1000)
clusters <- quickCluster(sce.pbmc)
sce.pbmc <- computeSumFactors(sce.pbmc, min.mean=0.1, cluster=clusters)
sce.pbmc <- normalize(sce.pbmc)
```

```{r Exploration5}
sce.pbmc
```

#### Select features for deminsionality reduction and clustering

```{r feature-selection}
fit.pbmc <- trendVar(sce.pbmc, use.spikes=FALSE)
dec.pbmc <- decomposeVar(fit=fit.pbmc)
o <- order(dec.pbmc$bio, decreasing=TRUE)
chosen.hvgs <- rownames(dec.pbmc)[head(o, 2000)]
```

#### Run Dimensionality Reduction: PCA, TSNE, and UMAP

```{r dimensionality-reduction}
set.seed(10000)
sce.pbmc <- runPCA(sce.pbmc, feature_set=chosen.hvgs, ncomponents=25,
    BSPARAM=BiocSingular::IrlbaParam())
set.seed(100000)
# BiocManager::install("Rtsne")
# install.packages("Rtsne")
sce.pbmc <- runTSNE(sce.pbmc, use_dimred="PCA")
set.seed(1000000)
# BiocManager::install("uwot")
# install.packages("uwot")
sce.pbmc <- runUMAP(sce.pbmc, use_dimred="PCA")
```

```{r Exploration6}
sce.pbmc
```

#### Run Clustering

```{r clustering}
g <- buildSNNGraph(sce.pbmc, k=10, use.dimred = 'PCA')
clust <- igraph::cluster_walktrap(g)$membership
sce.pbmc$cluster <- factor(clust)
```

```{r Exploration7}
sce.pbmc
```

```{r save, echo = FALSE, eval = FALSE, include=FALSE}
saveRDS("sce.pbmc_mid.rds")
```

```{r load, echo = FALSE, eval = FALSE, include=FALSE}
sce.pbmc <- readRDS("sce.pbmc_mid.rds")
```

## Before we start, what do we have?

```{r}
library(DittoSeq)
DEFAULT <- "sce.pbmc"
DBDimPlot("cluster", do.letter = FALSE)
```

We have a dataset that is broken into 12 clusters, but we do not know what each cluster represents.

## Determine Celltypes

Question: How do I run SingleR utilizing a known database?
The below code is obviously wrong.

```{r}
devtools::install_github("LTLA/SingleR")
library(SingleR)
singler.results <- SingleR(sce.pbmc, sce.pbmc, labels = sce.pbmc$cluster, method = "single")
singler.results
```
